### PR TITLE
Move visibility checks into speedy.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
This is part of https://github.com/digital-asset/daml/issues/9947

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 1]
```diff
@@ -76,34 +76,25 @@ object SResult {
...
-      committers: Set[Party],
-      // Callback.
-      // returns the next expression to evaluate.
-      // In case of failure the call back does not throw but set machine.ctrl to an SErrorDamlException
-      callback: SVisibleByKey => Unit,
-  ) extends SResult
-
-  sealed abstract class SVisibleByKey
-  object SVisibleByKey {
+  sealed abstract class SVisibility
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:79]: Maybe rename this to `SVisibleToStakeholders` or `SStakeholderAreReaders` (I’m sure you can come up with a less shitty name).

For contract ids it does not describe visibility properly and while we deprecate divulgence it’s still around for a while.

**Review by cocreature — APPROVED** [review]
Nice work, thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:79]: Done, thanks!

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 2]
```diff
@@ -87,6 +87,7 @@ class Engine(val config: EngineConfig = new EngineConfig(LanguageVersion.StableV
     */
   def submit(
       submitters: Set[Party],
+      readAs: Set[Party],
```
Comment (miklos-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:90]: Can you extend the code documentation above, please?

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala [hunk 3]
```diff
@@ -135,6 +135,7 @@ class ContractDiscriminatorFreshnessCheckSpec
     engine
       .submit(
         submitters = Set(alice),
+        readAs = Set.empty,
```
Comment (miklos-da) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala:138]: What does an empty set mean for `readAs`?

Comment (sofiafaro-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala:138]: It means there are no additional readers (besides the submitters).

Comment (cocreature) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala:138]: same semantics as on the ledger API

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 2 above)
Comment (sofiafaro-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:91]: Good idea, thanks!

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala (hunk 3 above)
Comment (miklos-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala:138]: Can you point me to the relevant documentation, please?

Comment (cocreature) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractDiscriminatorFreshnessCheckSpec.scala:138]: https://github.com/digital-asset/daml/blob/ffc88d52b438dce5703b1b20d7d82f6f3d63337b/ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto#L97

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 2 above)
Comment (miklos-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:91]: Thanks extending the code documentation!
If these are really additional readers (i.e., besides the submitters), I'd suggest naming the parameter accordingly. Though, I might be missing some context regarding why we chose `readAs` originally.

**Review by miklos-da — APPROVED** [review]

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:91]: I’d prefer using the same name everywhere. Using different names for the same thing across our codebase is just confusing. As for why we have those names, I believe we tried to match the names we used for authorization when we added multitparty submissions.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 4]
```diff
@@ -76,34 +76,25 @@ object SResult {
...
-      // Callback.
-      // returns the next expression to evaluate.
-      // In case of failure the call back does not throw but set machine.ctrl to an SErrorDamlException
-      callback: SVisibleByKey => Unit,
-  ) extends SResult
-
-  sealed abstract class SVisibleByKey
-  object SVisibleByKey {
+  sealed abstract class SVisibleToStakeholders
+  object SVisibleToStakeholders {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:80]: A bit weird  to have the same name for the base type and the case object.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 5]
```diff
@@ -120,7 +122,11 @@ private[lf] object Speedy {
       var globalDiscriminators: Set[crypto.Hash],
       var cachedContracts: Map[V.ContractId, CachedContract],
   ) extends LedgerMode {
-
+    private[lf] val visibleToStakeholders: Set[Party] => SVisibleToStakeholders =
+      if (validating) { _ => SVisibleToStakeholders.Visible }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:126]: Naive question, why do we skip the check on validation ?

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 5 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:126]: Visibility checks are a submission-only concept. They exist primarily to protect parties on the same participant from each other.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala (hunk 4 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:80]: It's not a case object.

I didn't change anything here though, only renamed the existing type.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:80]: My mistake.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 2 above)
Comment (rautenrieth-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:91]: The `readAs` and `actAs` terms were first used when we added ledger API authorization, for parties the bearer of the auth token is allowed to use for read-only (e.g., transaction service) and state modifying (e.g., command service) ledger API queries, respectively. At this time, the concept of multi-party submissions did not exist and `readAs` had nothing to do with contract visibility during command interpretation.

The terms then made it into the ledger API and the engine when we added multi-party submissions, where `readAs` means "parties that can be used to look up contracts/keys, in addition to the submitters". Using a `readAs` party for multi-party submissions requires a corresponding `readAs` claim in the auth token.

Note also the following details:
- The `readAs`/ `actAs` parties from auth tokens are a superset of the parties used in ledger API commands (you don't have to use all parties you are allowed to)
- The `readAs`/ `actAs` parties from ledger API commands are a superset of the minimal sets of parties required to successfully submit the command (you can submit a command as Alice+Bob, even though Alice alone would be enough to authorize the command in the Daml model)
- The ledger API server normalizes `readAs` parties by removing `actAs` parties (as `actAs` implies read access)

`submitters` and `actAs` are the probably the same thing.

**Pull request merged**